app-creator/src/redux/types/attributes.ts

```diff
@@ -0,0 +1,283 @@
+ */
+import { InputType } from './enums';
+
+export type SharedAttributes =
+  | 'height'
+  | 'width'
+  | 'padding'
+  | 'margin'
+  | 'color'
+  | 'backgroundColor'
+  | 'borderWidth'
+  | 'borderType'
+  | 'borderColor'
+  | 'fontSize'
+  | 'fontWeight'
+  | 'fontFamily'
+  | 'textAlign'
+  | 'whiteSpace'
+  | 'shown';
+
+export interface AttributeMetaData {
+  [key: string]: {
+    value: string;
+    type: InputType;
+    items?: string[];
+  };
+}
+
+export interface DefaultAttributesType {
+  [key: string]: string;
+}
+
+export const sharedAttributes: AttributeMetaData = {
+  height: {
+    value: '',
+    type: InputType.number,
+  },
+  width: {
+    value: '',
+    type: InputType.number,
+  },
+  padding: {
+    value: '0',
+    type: InputType.number,
+  },
+  margin: {
+    value: '8',
+    type: InputType.number,
+  },
+  color: {
+    value: 'black',
+    type: InputType.color,
+  },
+  backgroundColor: {
+    value: 'none',
+    type: InputType.color,
+  },
+  borderWidth: {
+    value: '0',
+    type: InputType.number,
+  },
+  borderType: {
+    value: 'solid',
+    type: InputType.select,
+    items: ['solid', 'dashed'],
+  },
+  borderColor: {
+    value: 'black',
+    type: InputType.color,
+  },
+  fontSize: {
+    value: '12',
+    type: InputType.number,
+  },
+  fontWeight: {
+    value: '300',
+    type: InputType.select,
+    items: ['100', '200', '300', '400', '500', '600', '700', '800', '900'],
+  },
+  fontFamily: {
+    value: 'Roboto',
+    type: InputType.select,
+    items: [
+      'Arial',
+      'Arial Black',
+      'Bookman',
+      'Candara',
+      'Comic Sans MS',
+      'Courier',
+      'Courier New',
+      'Garamond',
+      'Georgia',
+      'Impact',
+      'Open Sans',
+      'Palatino',
+      'Roboto',
+      'Times',
+      'Times New Roman',
+      'Verdana',
+    ],
+  },
+  textAlign: {
+    value: 'left',
+    type: InputType.select,
+    items: ['left', 'center', 'right', 'justify'],
+  },
+  whiteSpace: {
+    value: 'normal',
+    type: InputType.select,
+    items: ['normal', 'pre'],
+  },
+  shown: {
+    value: 'true',
+    type: InputType.select,
+    items: ['true', 'false'],
+  },
+};
+
+export const DEFAULT_SHARED_ATTRIBUTES: DefaultAttributesType = Object.keys(
+  sharedAttributes
+).reduce((attributes, key) => {
+  if (sharedAttributes[key].type === InputType.number) {
+    return { ...attributes, [key]: sharedAttributes[key].value + 'px' };
+  }
+  return { ...attributes, [key]: sharedAttributes[key].value };
+}, {});
+
+export const textboxAttributes: AttributeMetaData = {
+  value: {
+    value: '',
+    type: InputType.text,
+  },
+  placeholder: {
+    value: 'Enter text',
+    type: InputType.text,
+  },
+};
+
+export const DEFAULT_TEXTBOX_ATTRIBUTES: DefaultAttributesType = Object.keys(
+  textboxAttributes
+).reduce(
+  (attributes, key) => ({ ...attributes, [key]: textboxAttributes[key].value }),
+  {}
+);
```

Comment:
Nice ES6ing here. It's succinct and clean, but requires careful understanding of ES6 features to get right. In fact, the object spread operator is still experimental ES2018! https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#Spread_in_object_literals

I would suggest using Object.assign instead in order to keep within ES6 features and to be less cryptic.

(Also for above and below)

Reply:
What about `reducer.ts`? We are using it often there.  

```Typescript
switch (action.type) {
    case SET_DRAGGING_WIDGET:
      return {
        ...state,
        ...action.payload,
      };
    case SET_EDITING_WIDGET:
      return {
        ...state,
        element: action.payload.element,
        eventType: action.payload.eventType,
        selectedTab: action.payload.openAttributesTab
          ? Tab.attributes
          : state.selectedTab,
      };
    case SET_SELECTED_TAB:
      return {
        ...state,
        ...action.payload,
      };
     ...
```